model_analyzer/record/record_aggregator.py

```diff
@@ -0,0 +1,159 @@
+            imposing the filter criteria on.
+
+        filters : list of callables
+            conditions that determine whether
+            a given record should be returned.
+            If no filters specified, all records
+            of types specified by headers will be
+            returned
+
+        Returns
+        -------
+        Dict of List of Records
+            keys are record headers, values
+            are all records of that type
+        """
+        if not headers and not filters:
+            return self._records
+        if headers and not filters:
+            try:
+                return {k: self._records[k] for k in headers}
+            except KeyError as k:
+                raise TritonModelAnalyzerException(
+                    f"Record type '{k}' not found in this RecordAggregator")
+        if filters and not headers:
+            raise TritonModelAnalyzerException(
+                "Must specify the headers corresponding to each filter criterion."
+            )
+        if len(headers) != len(filters):
+            raise TritonModelAnalyzerException(
+                "Must specify the same number of headers as filter criteria.")
+
+        # Remove records that do not satisfy criteria
+        filtered_records = defaultdict(list)
+        for h, f in zip(headers, filters):
+            filtered_records[h] = [
+                record for record in self._records[h] if f(record.value())
+            ]
+
+        return filtered_records
+
+    def headers(self):
+        """
+        Returns
+        -------
+        list of str
+            a list of the types of records in this
+            RecordAgrregator
+        """
+        return list(self._records.keys())
+
+    def total(self, header=None):
+        """
+        Get the total number of records in
+        the RecordAggregator
+
+        Parameters
+        ----------
+        header : str
+            The type of records to count,
+            if None, count all types
+
+        Returns
+        -------
+        int
+            number of records in
+            the RecordAggregator
+        """
+        if header:
+            if header not in self._records:
+                raise TritonModelAnalyzerException(
+                    f"Record type '{header}' not found in this RecordAggregator"
+                )
+            return len(self._records[header])
+        return sum(len(self._records[k]) for k in self._records)
+
+    def aggregate(self, headers, reduce_func):
+        """
+        Parameters
+        ----------
+        headers : str
+            The type of records to aggregate
+
+        reduce_func : callable
+            takes as input a list of values
+            and returns one
+
+        Returns
+        -------
+        dict
+            keys are requested headers
+            and values are the aggregated values
+        """
+
+        return {
+            k: reduce_func([record.value() for record in self._records[k]])
```

Comment:
I am not a big fan of this style of programming as it obfuscates what is actually going on. Pull it out of the return and create an actual object the is returned. The compiler will optimize out any unnecessary variables.

Reply:
Will change it for readability, but I think list/dictionary comprehensions are the most pythonic way to write this. They read like mathematical objects, and are something like 50% faster than the corresponding for loop semantics. Most python libraries are full of them.